### ISSUE TITLE
Aggregated logs generated by --bandwidth-log option are capped at 1024 lines

### ISSUE THREAD

**Issue by pjakuszew** [axboe/fio#1113]
Hello,

I have discovered that `--bandwidth-log` command line option produces incomplete logs on longer test run times. This was discovered on fio-3.23 built from source on Ubuntu 18.04.4.

Used command:
```
./fio-3.23/fio --filename=/dev/nvme1n1p1 --direct=1 --rw=randrw --refill_buffers --norandommap --randrepeat=0 --ioengine=libaio --bs=4k --rwmixread=50 --iodepth=16 --numjobs=16 --runtime=24h --time_based --group_reporting --name=4kxxxtest --output=output.log --bandwidth-log
```

I have tried to determine the cause of bug but was unable to find it, so far.

I suspect that `io_logs` struct containing the aggregated logs does not have its entry count incremented properly, which makes the following statement to fail and the resulting `aggr-*_bw.log` files are capped at 1024 lines (current value of `DEF_LOG_ENTRIES` defined in iolog.h):
https://github.com/axboe/fio/blob/fio-3.23/stat.c#L2431-L2441
```
	/*
	 * Cap the size at MAX_LOG_ENTRIES, so we don't keep doubling
	 * forever
	 */
	if (!iolog->cur_log_max)
		new_samples = DEF_LOG_ENTRIES;
	else {
		new_samples = iolog->cur_log_max * 2;
		if (new_samples > MAX_LOG_ENTRIES)
			new_samples = MAX_LOG_ENTRIES;
	}
```

This issue does not happen when `write_bw_log` option is used; per-job bandwidth logs contain full metrics. I tried to compare the sample append differences between `write_bw_log` and `bandwidth-log` options, but frankly I'm not that familiar with C to understand it.

I have temporarily mitigated this in my test case by changing `DEF_LOG_ENTRIES` to `MAX_LOG_ENTRIES` at https://github.com/axboe/fio/blob/fio-3.23/stat.c#L2436 but I would appreciate any tips about the proper way of fixing this issue.

**Comment by pjakuszew** [comment]
I have added fio output and bandwidth logs in attachments, they were produced from following command on my computer:
```
../fio/fio --filename=fiotest.bin --direct=1 --rw=randrw --refill_buffers --norandommap --randrepeat=0 --ioengine=libaio --bs=4k --rwmixread=50 --iodepth=16 --numjobs=4 --runtime=13m --time_based --group_reporting --name=4kxxxtest --bandwidth-log --write_bw_log=perjob_test --output=output.log
```

I will not attach the whole per-job log file because its size is 250MB, instead I'm attaching last 10 lines.

[agg-read_bw.log](https://github.com/axboe/fio/files/5465470/agg-read_bw.log)
[agg-trim_bw.log](https://github.com/axboe/fio/files/5465471/agg-trim_bw.log)
[agg-write_bw.log](https://github.com/axboe/fio/files/5465472/agg-write_bw.log)
[last_10_lines__perjob_test_bw.1.log](https://github.com/axboe/fio/files/5465473/last_10_lines__perjob_test_bw.1.log)
[output.log](https://github.com/axboe/fio/files/5465474/output.log)

**Comment by sitsofe** [comment]
Seems reproducible:
```
$ fio --name=aggloglimitedto1024 --ioengine=null --time_based=1 --runtime=10s --size=100M --bandwidth-log --numjobs=1 --bwavgtime=5
aggloglimitedto1024: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=null, iodepth=1
fio-3.23-50-g8c17
Starting 1 process
Jobs: 1 (f=1): [R(1)][100.0%][r=7688MiB/s][r=1968k IOPS][eta 00m:00s]
aggloglimitedto1024: (groupid=0, jobs=1): err= 0: pid=9021: Sat Oct 31 16:14:37 2020
  read: IOPS=1958k, BW=7649MiB/s (8021MB/s)(74.7GiB/10000msec)
[...]
     issued rwts: total=19582430,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1
$ wc -l *.log
    1024 agg-read_bw.log
    1024 agg-trim_bw.log
    1024 agg-write_bw.log
    3072 total
```

I would have guessed the log would be able to grow to `MAX_LOG_ENTRIES` (from https://github.com/axboe/fio/blob/8c17a6248bd227a9f3cc3da52bd1cb922dc6cf81/iolog.h#L63-L64 ):
```
#define DEF_LOG_ENTRIES         1024
#define MAX_LOG_ENTRIES         (1024 * DEF_LOG_ENTRIES)
```

**Comment by sitsofe** [comment]
@pjakuszew: I'd guess something like the following would do what you want but I don't know if it's a good idea to do every time aggregate I/O has a sample...
```diff
diff --git a/eta.c b/eta.c
index d1c9449f..eaad1ac9 100644
--- a/eta.c
+++ b/eta.c
@@ -508,6 +508,7 @@ bool calc_thread_status(struct jobs_eta *je, int force)
 				je->rate);
 		memcpy(&rate_prev_time, &now, sizeof(now));
 		for_each_rw_ddir(ddir) {
+			regrow_agg_logs();
 			add_agg_sample(sample_val(je->rate[ddir]), ddir, 0, 0);
 		}
 	}
diff --git a/iolog.h b/iolog.h
index 981081f9..9e382cc0 100644
--- a/iolog.h
+++ b/iolog.h
@@ -182,6 +182,7 @@ static inline struct io_sample *__get_sample(void *samples, int log_offset,
 struct io_logs *iolog_cur_log(struct io_log *);
 uint64_t iolog_nr_samples(struct io_log *);
 void regrow_logs(struct thread_data *);
+void regrow_agg_logs(void);
 
 static inline struct io_sample *get_sample(struct io_log *iolog,
 					   struct io_logs *cur_log,
diff --git a/stat.c b/stat.c
index eb40bd7f..d42586e7 100644
--- a/stat.c
+++ b/stat.c
@@ -2536,6 +2536,14 @@ void regrow_logs(struct thread_data *td)
 	td->flags &= ~TD_F_REGROW_LOGS;
 }
 
+void regrow_agg_logs(void)
+{
+	enum fio_ddir ddir;
+
+	for (ddir = 0; ddir < DDIR_RWDIR_CNT; ddir++)
+		regrow_log(agg_io_log[ddir]);
+}
+
 static struct io_logs *get_cur_log(struct io_log *iolog)
 {
 	struct io_logs *cur_log;
```

**Comment by pjakuszew** [comment]
Thanks for taking a look at this, @sitsofe - indeed, I noticed that `bandwidth-log` option is implemented a bit differently than the rest of logging options like `write_bw_log` job parameter.

We prefer using `bandwidth-log` in our testcases because we attempt to simulate 24h-long heavy workloads on different drives to compare performance between disks made by different vendors. Because we use a bigger amount of `numjobs`, this option creates a lot of log files which are rather difficult to parse later with fio2gnuplot due to their size - it takes really long time to create graphs out of them.

Side note: I have also noticed that aggregated logs do not contain block size used in test...
